neo/rawio/bci2000rawio.py

```diff
@@ -0,0 +1,346 @@
+# -*- coding: utf-8 -*-
+"""
+BCI2000RawIO is a class to read BCI2000 .dat files.
+https://www.bci2000.org/mediawiki/index.php/Technical_Reference:BCI2000_File_Format
+"""
+from __future__ import unicode_literals, print_function, division, absolute_import
+
+from .baserawio import (BaseRawIO, _signal_channel_dtype, _unit_channel_dtype, 
+        _event_channel_dtype)
+
+import numpy as np
+import re
+import urllib
+
+class BCI2000RawIO(BaseRawIO):
+    """
+    Class for reading data from a BCI2000 .dat file, either version 1.0 or 1.1
+    """
+    extensions = ['dat']
+    rawmode = 'one-file'
+
+    def __init__(self, filename=''):
+        BaseRawIO.__init__(self)
+        self.filename = filename
+        self._my_events = None
+
+    def _source_name(self):
+        return self.filename
+    
+    def _parse_header(self):
+        file_info, state_defs, param_defs = parse_bci2000_header(self.filename)
+
+        self.header = {}
+        self.header['nb_block'] = 1
+        self.header['nb_segment'] = [1]
+
+        sig_channels = []
+        for chan_ix in range(file_info['SourceCh']):
+            ch_name = param_defs['ChannelNames']['value'][chan_ix] if 'ChannelNames' in param_defs else 'ch' + str(chan_ix)
+            chan_id = chan_ix + 1
+            sr = param_defs['SamplingRate']['value'] #Hz
+            dtype = file_info['DataFormat']
+            units = 'uV'
+            gain = param_defs['SourceChGain']['value'][chan_ix]
+            offset = param_defs['SourceChOffset']['value'][chan_ix]
+            group_id = 0 
+            sig_channels.append((ch_name, chan_id, sr, dtype,  units, gain,offset, group_id))
+        self.header['signal_channels'] = np.array(sig_channels, dtype=_signal_channel_dtype)
+
+        self.header['unit_channels'] = np.array([], dtype=_unit_channel_dtype)
+        
+        # creating event channel for each state variable
+        event_channels = []
+        for st_ix, st_tup in enumerate(state_defs):
+            event_channels.append((st_tup[0], 'ev_' + str(st_ix), 'event'))
```

Comment:
In your code it seems that here it should be 'epoch' and not 'event' because duration is not None.

Reply:
Almost all state changes in BCI2000 are stimulus-related; the state will go from 0 to some value when it turns on, then back to 0 when it turns off. Sometimes the transition back to 0 is nonsense, e.g. when the state change triggers an animation in a game engine, the state will change back to 0 after a minimum duration of one data block, but the animation might continue for some unknowable time. The duration is often not useful, but sometimes it is, so I would like to hold on to the 'duration' information for experts who need it.

Another issue is that I predict calling these 'epochs' will introduce confusion due to context. BCI2000 is more popular in EEG and ECoG than in microelectrode neurophys, so many people that might use Neo to import BCI2000 will be coming from those backgrounds and might already be familiar with EEGLAB. In EEGLAB, the BCI2000 importer (to which I contribute) imports the zero-to-nonzero state transitions as events, and the word 'epoch' is strongly associated with the concept of an entire 'trial'.

I don't mind calling them epochs instead but that's easy because I'm aware of the issue. Or do you think it would be better to keep them as events but maybe give each an annotation holding the duration, for the expert users?

Reply:
I am aware of that confusion for people from eeg.

'event' or 'epoch' is not the name it refer to the type.
_event_channel_dtype = [('name','U64'),('id','U64'),('type', 'S5')]   # type = epoch or event
It is part of tha API to tell neo.io if this event channel is in fact a neo.Event (no duration) or a neo.Epoch (with duration).
So in your case, if you put 'event' this will never be translate to neo.Epoch and so loose the duration at neo.io level.

Maybe you can create 2 channel which duplicate artificially the same triggers.
One is 'event' and will be a neo.Event and one is 'epoch' and will be a neo.Epoch at neo.io.
Depending of chanel_index you can set duration is None or not.

Reply:
Thanks for the clarification about data types. I've been working almost exclusively at the rawio level.

Maybe I'm overthinking this. I think I will try keep the stimulus-off transitions as events, then the user can calculate their own duration if they need it, or they can discard/ignore these events if they don't.